src/modding/ModLoader.cs

```diff
@@ -0,0 +1,278 @@
+            {
+                var info = ModManager.LoadModInfo(modsFolder);
+
+                if (info == null)
+                {
+                    GD.PrintErr("Failed to load info for mod \"", name, "\" from: ", modsFolder);
+                    return null;
+                }
+
+                return new FullModDetails(name) { Folder = modsFolder, Info = info };
+            }
+        }
+
+        if (failureIsError)
+            GD.PrintErr("No folder found for mod: ", name);
+        return null;
+    }
+
+    public override void _Ready()
+    {
+        base._Ready();
+
+        if (modInterface != null)
+            throw new InvalidOperationException("ModInterface has been created already");
+
+        modInterface = new ModInterface(GetTree());
+
+        LoadMods();
+    }
+
+    public override void _Process(float delta)
+    {
+        base._Process(delta);
+
+        if (firstExecute)
+        {
+            GD.Print("Loading mod Nodes into the scene tree");
+
+            foreach (var tuple in loadedModAssemblies)
+            {
+                RunCodeModFirstRunCallbacks(tuple.Value);
+            }
+
+            firstExecute = false;
+        }
+    }
+
+    public void LoadMods()
+    {
+        var newMods = Settings.Instance.EnabledMods.Value.ToHashSet();
+
+        foreach (var unload in loadedMods.Where(m => !newMods.Contains(m)).ToList())
+        {
+            GD.Print("Unloading mod: ", unload);
+            UnLoadMod(unload);
+            loadedMods.Remove(unload);
+        }
+
+        if (newMods.Count < 1)
+            return;
+
+        foreach (var load in newMods.Except(loadedMods).ToList())
+        {
+            GD.Print("Loading mod: ", load);
+            LoadMod(load);
+            loadedMods.Add(load);
+        }
+
+        GD.Print("Mod loading finished");
+    }
+
+    private void LoadMod(string name)
+    {
+        var info = LoadModInfo(name);
+
+        if (info == null)
+        {
+            GD.PrintErr("Can't load mod due to failed info reading: ", name);
+            return;
+        }
+
+        bool loadedSomething = false;
+
+        if (!string.IsNullOrEmpty(info.Info.PckToLoad))
+        {
+            LoadPckFile(Path.Combine(info.Folder, info.Info.PckToLoad));
+            loadedSomething = true;
+        }
+
+        if (!string.IsNullOrEmpty(info.Info.ModAssembly))
+        {
+            Assembly modAssembly;
+            try
+            {
+                modAssembly = LoadCodeAssembly(Path.Combine(info.Folder, info.Info.ModAssembly));
+            }
+            catch (Exception e)
+            {
+                GD.PrintErr("Could not load mod assembly due to exception: ", e);
+                return;
+            }
+
+            if (!CreateModInstance(name, info, modAssembly))
+                return;
+
+            loadedSomething = true;
+        }
+
+        if (!loadedSomething)
+        {
+            GD.Print("A mod contained no loadable resources");
```

Comment:
```suggestion
            GD.Print($"The mod ${name} contained no loadable resources");
```

Reply:
A bit weird to have one string in the file use `$` whereas other prints don't use.

Reply:
I think I skipped the name here as the upper level method prints the mod's name anyway, so this is not needed.